storage: TestReplicateReAddAfterDown

The following test appears to have failed:

[#8532](https://circleci.com/gh/cockroachdb/cockroach/8532):

```
E1021 16:55:05.474847 945 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
E1021 16:55:06.475309 945 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
I1021 16:55:07.476165 945 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
I1021 16:55:07.476329 945 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
I1021 16:55:07.476437 945 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- FAIL: TestReplicateReAddAfterDown (3.25s)
	testing.go:157: storage/client_raft_test.go:1405: condition failed to evaluate within 3s: storage/client_test.go:511: expected [16 16 16], got [16 16 0]
=== RUN   TestSetupRangeTree
I1021 16:55:07.477039 945 gossip/gossip.go:175  gossiping node descriptor node_id:1 address:<network_field:"localhost" address_field:"1" > attrs:<> 
I1021 16:55:07.477787 945 kv/range_cache.go:128  lookup range descriptor: key="\x00range-tree-root"
I1021 16:55:07.477814 945 kv/range_cache.go:128  lookup range descriptor: key="\x00\x00meta2\x00range-tree-root"
I1021 16:55:07.478015 945 kv/range_cache.go:189  adding descriptor: key="\x00\x00meta2\xff\xff" desc=range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I1021 16:55:07.478134 945 kv/range_cache.go:189  adding descriptor: key="\x00\x00meta2\xff\xff" desc=range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I1021 16:55:07.478189 945 kv/range_cache.go:313  clearing overlapping descriptor: key="\x00\x00meta2\xff\xff" desc=range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I1021 16:55:07.478349 945 multiraft/multiraft.go:524  node 1 starting
I1021 16:55:07.478462 945 raft/raft.go:406  [group 1] 1 became follower at term 5
--

I1021 16:55:09.238658 945 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestStoreRangeSystemSplits (0.23s)
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.14s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	14.120s
=== RUN   TestBatchBasics
I1021 16:54:46.615487 910 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I1021 16:54:46.616233 910 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN   TestBatchMerge
I1021 16:54:46.618231 910 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

https://circleci.com/gh/tschottdorf/cockroach/1520

>
I1022 14:57:50.001559 975 multiraft/multiraft.go:778  node 3: got message for unknown group 1; creating it
I1022 14:57:50.001692 975 raft/raft.go:409  [group 1] 4 became follower at term 6
I1022 14:57:50.001758 975 raft/raft.go:218  [group 1] newRaft 4 [peers: [1,2,3], term: 6, commit: 17, applied: 17, lastindex: 17, lastterm: 6]
I1022 14:57:50.006880 975 util/tracer/tracer.go:138  Name                 Origin  Ts               Dur        Desc                     File
c525870006369319.41          14:57:50.006419  308.914µs  sending batch            kv/txn_coord_sender.go:306
c525870006369319.41          14:57:50.006455  41.974µs   ·meta descriptor lookup  kv/dist_sender.go:501
c525870006369319.41          14:57:50.006506  218.06µs   ·sending RPC             kv/dist_sender.go:421

>
I1022 14:57:50.006933 975 storage/replica_gc_queue.go:125  destroying local data from range 1
I1022 14:57:50.008722 975 multiraft/multiraft.go:778  node 3: got message for unknown group 1; creating it
W1022 14:57:50.008788 975 multiraft/multiraft.go:782  Error creating group 1 (in response to incoming message): raft group deleted
E1022 14:57:50.012464 975 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
E1022 14:57:50.046618 975 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
E1022 14:57:50.114098 975 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
E1022 14:57:50.248693 975 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
E1022 14:57:50.517442 975 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
E1022 14:57:51.054675 975 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
E1022 14:57:52.055020 975 storage/client_test.go:497  error reading "a" from engine 2: value type is not INT: UNKNOWN
I1022 14:57:53.055838 975 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
I1022 14:57:53.055951 975 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
I1022 14:57:53.056009 975 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
	testing.go:157: storage/client_raft_test.go:1406: condition failed to evaluate within 3s: storage/client_test.go:511: expected [16 16 16], got [16 16 0]

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -1395,8 +1395,13 @@ func TestReplicateReAddAfterDown(t *testing.T) {
 	}
 	mtc.waitForValues(roachpb.Key("a"), 3*time.Second, []int64{16, 16, 5})
 
-	// Bring it back up and re-add the range.
-	// This re-uses the existing replica but changes its replica ID.
+	// Bring it back up and re-add the range. There is a race when the
+	// store applies its removal and re-addition back to back: the
+	// replica may or may not have (asynchronously) garbage collected
+	// its data in between. Whether the existing data is reused or the
+	// replica gets recreated, the replica ID is changed by this
+	// process. An ill-timed GC has been known to cause bugs including
+	// https://github.com/cockroachdb/cockroach/issues/2873.
 	mtc.restartStore(2)
 	mtc.replicateRange(raftID, 0, 2)
 

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -233,8 +233,12 @@ func (r *Replica) String() string {
 }
 
 // Destroy cleans up all data associated with this range, leaving a tombstone.
-func (r *Replica) Destroy() error {
+func (r *Replica) Destroy(origDesc roachpb.RangeDescriptor) error {
 	desc := r.Desc()
+	if _, rd := desc.FindReplica(r.store.StoreID()); rd != nil && rd.ReplicaID >= origDesc.NextReplicaID {
+		return util.Errorf("cannot destroy replica %s; replica ID has changed (%s >= %s)",
+			r, rd.ReplicaID, origDesc.NextReplicaID)
+	}
 	iter := newReplicaDataIterator(desc, r.store.Engine())
 	defer iter.Close()
 	batch := r.store.Engine().NewBatch()

--- a/storage/replica_data_iter_test.go
+++ b/storage/replica_data_iter_test.go
@@ -188,7 +188,7 @@ func TestReplicaDataIterator(t *testing.T) {
 	}
 
 	// Destroy range and verify that its data has been completely cleared.
-	if err := tc.rng.Destroy(); err != nil {
+	if err := tc.rng.Destroy(*tc.rng.Desc()); err != nil {
 		t.Fatal(err)
 	}
 	iter = newReplicaDataIterator(tc.rng.Desc(), tc.rng.store.Engine())

--- a/storage/replica_gc_queue.go
+++ b/storage/replica_gc_queue.go
@@ -122,7 +122,7 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
 		if log.V(1) {
 			log.Infof("destroying local data from range %d", desc.RangeID)
 		}
-		if err := rng.store.RemoveReplica(rng); err != nil {
+		if err := rng.store.RemoveReplica(rng, replyDesc); err != nil {
 			return err
 		}
 
@@ -143,7 +143,7 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
 		if err := rng.Destroy(); err != nil {
 			return err
 		}
-	} else if desc.RangeID != desc.RangeID {
+	} else if desc.RangeID != replyDesc.RangeID {
 		// If we get a different  range ID back, then the range has been merged
 		// away. But currentMember is true, so we are still a member of the
 		// subsuming range. Shut down raft processing for the former range
@@ -151,7 +151,7 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
 		if log.V(1) {
 			log.Infof("removing merged range %d", desc.RangeID)
 		}
-		if err := rng.store.RemoveReplica(rng); err != nil {
+		if err := rng.store.RemoveReplica(rng, replyDesc); err != nil {
 			return err
 		}
 

--- a/storage/replica_gc_queue.go
+++ b/storage/replica_gc_queue.go
@@ -84,6 +84,9 @@ func (*replicaGCQueue) shouldQueue(now roachpb.Timestamp, rng *Replica,
 // process performs a consistent lookup on the range descriptor to see if we are
 // still a member of the range.
 func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.SystemConfig) error {
+	// Note that the Replicas field of desc is probably out of date, so
+	// we should only use `desc` for its static fields like RangeID and
+	// StartKey (and avoid rng.GetReplica() for the same reason).
 	desc := rng.Desc()
 
 	// Calls to RangeLookup typically use inconsistent reads, but we

--- a/storage/replica_gc_queue.go
+++ b/storage/replica_gc_queue.go
@@ -109,12 +109,11 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
 
 	replyDesc := reply.Ranges[0]
 	currentMember := false
-	if me := rng.GetReplica(); me != nil {
-		for _, rep := range replyDesc.Replicas {
-			if rep.StoreID == me.StoreID {
-				currentMember = true
-				break
-			}
+	storeID := rng.store.StoreID()
+	for _, rep := range replyDesc.Replicas {
+		if rep.StoreID == storeID {
+			currentMember = true
+			break
 		}
 	}
 

--- a/storage/replica_gc_queue.go
+++ b/storage/replica_gc_queue.go
@@ -140,7 +140,7 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
 			return nil
 		}
 
-		if err := rng.Destroy(); err != nil {
+		if err := rng.Destroy(replyDesc); err != nil {
 			return err
 		}
 	} else if desc.RangeID != replyDesc.RangeID {

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -64,6 +64,7 @@ func testRangeDescriptor() *roachpb.RangeDescriptor {
 				StoreID:   1,
 			},
 		},
+		NextReplicaID: 2,
 	}
 }
 
@@ -3412,4 +3413,34 @@ func TestReplicaLoadSystemDBSpanIntent(t *testing.T) {
 		}
 		return nil
 	})
+}
+
+func TestReplicaDestroy(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	tc := testContext{}
+	tc.Start(t)
+	defer tc.Stop()
+
+	rep, err := tc.store.GetReplica(1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// First try and fail with an outdated descriptor.
+	origDesc := rep.Desc()
+	newDesc := proto.Clone(origDesc).(*roachpb.RangeDescriptor)
+	_, newRep := newDesc.FindReplica(tc.store.StoreID())
+	newRep.ReplicaID++
+	newDesc.NextReplicaID++
+	if err := rep.setDesc(newDesc); err != nil {
+		t.Fatal(err)
+	}
+	if err := rep.Destroy(*origDesc); !testutils.IsError(err, "replica ID has changed") {
+		t.Fatalf("expected error 'replica ID has changed' but got %s", err)
+	}
+
+	// Now try a fresh descriptor and succeed.
+	if err := rep.Destroy(*rep.Desc()); err != nil {
+		t.Fatal(err)
+	}
 }

--- a/storage/store.go
+++ b/storage/store.go
@@ -1005,7 +1005,7 @@ func (s *Store) MergeRange(subsumingRng *Replica, updatedEndKey roachpb.RKey, su
 
 	// Remove and destroy the subsumed range. Note that we are on the
 	// processRaft goroutine so we can call removeReplicaImpl directly.
-	if err := s.removeReplicaImpl(subsumedRng); err != nil {
+	if err := s.removeReplicaImpl(subsumedRng, *subsumedDesc); err != nil {
 		return util.Errorf("cannot remove range %s", err)
 	}
 
@@ -1068,22 +1068,30 @@ func (s *Store) addReplicaToRangeMap(rng *Replica) error {
 }
 
 type removeReplicaOp struct {
-	rep *Replica
-	ch  chan<- error
+	rep      *Replica
+	origDesc roachpb.RangeDescriptor
+	ch       chan<- error
 }
 
 // RemoveReplica removes the replica from the store's replica map and from
-// the sorted replicasByKey btree.
-func (s *Store) RemoveReplica(rep *Replica) error {
+// the sorted replicasByKey btree. The version of the replica descriptor that
+// was used to make the removal decision is passed in, and the removal is
+// aborted if the replica ID has changed since then.
+func (s *Store) RemoveReplica(rep *Replica, origDesc roachpb.RangeDescriptor) error {
 
 	ch := make(chan error)
-	s.removeReplicaChan <- removeReplicaOp{rep, ch}
+	s.removeReplicaChan <- removeReplicaOp{rep, origDesc, ch}
 	return <-ch
 }
 
 // removeReplicaImpl runs on the processRaft goroutine.
-func (s *Store) removeReplicaImpl(rep *Replica) error {
-	rangeID := rep.Desc().RangeID
+func (s *Store) removeReplicaImpl(rep *Replica, origDesc roachpb.RangeDescriptor) error {
+	desc := rep.Desc()
+	rangeID := desc.RangeID
+	if _, rd := desc.FindReplica(s.StoreID()); rd != nil && rd.ReplicaID >= origDesc.NextReplicaID {
+		return util.Errorf("cannot remove replica %s; replica ID has changed (%s >= %s)",
+			rep, rd.ReplicaID, origDesc.NextReplicaID)
+	}
 
 	// Silence the Replica. This clears all outstanding commands and makes
 	// sure that whatever else slips in winds up with a RangeNotFoundError.
@@ -1563,7 +1571,7 @@ func (s *Store) processRaft() {
 				}
 
 			case op := <-s.removeReplicaChan:
-				op.ch <- s.removeReplicaImpl(op.rep)
+				op.ch <- s.removeReplicaImpl(op.rep, op.origDesc)
 
 			case op := <-s.proposeChan:
 				op.ch <- s.proposeRaftCommandImpl(op.idKey, op.cmd)

--- a/storage/store_test.go
+++ b/storage/store_test.go
@@ -270,7 +270,7 @@ func TestStoreAddRemoveRanges(t *testing.T) {
 		t.Error(err)
 	}
 	// Remove range 1.
-	if err := store.RemoveReplica(rng1); err != nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc()); err != nil {
 		t.Error(err)
 	}
 	// Create a new range (id=2).
@@ -287,7 +287,7 @@ func TestStoreAddRemoveRanges(t *testing.T) {
 		t.Fatalf("expected rangeAlreadyExists error; got %s", err)
 	}
 	// Try to remove range 1 again.
-	if err := store.RemoveReplica(rng1); err == nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc()); err == nil {
 		t.Fatal("expected error re-removing same range")
 	}
 	// Try to add a range with previously-used (but now removed) ID.
@@ -326,6 +326,33 @@ func TestStoreAddRemoveRanges(t *testing.T) {
 	}
 }
 
+func TestStoreRemoveReplicaOldDescriptor(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	store, _, stopper := createTestStore(t)
+	defer stopper.Stop()
+
+	rng1, err := store.GetReplica(1)
+	if err != nil {
+		t.Fatal(err)
+	}
+	origDesc := rng1.Desc()
+	newDesc := proto.Clone(origDesc).(*roachpb.RangeDescriptor)
+	_, newRep := newDesc.FindReplica(store.StoreID())
+	newRep.ReplicaID++
+	newDesc.NextReplicaID++
+	if err := rng1.setDesc(newDesc); err != nil {
+		t.Fatal(err)
+	}
+	if err := store.RemoveReplica(rng1, *origDesc); !testutils.IsError(err, "replica ID has changed") {
+		t.Fatalf("expected error 'replica ID has changed' but got %s", err)
+	}
+
+	// Now try the latest descriptor and succeed.
+	if err := store.RemoveReplica(rng1, *rng1.Desc()); err != nil {
+		t.Fatal(err)
+	}
+}
+
 func TestStoreRangeSet(t *testing.T) {
 	defer leaktest.AfterTest(t)
 	store, _, stopper := createTestStore(t)
@@ -336,7 +363,7 @@ func TestStoreRangeSet(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := store.RemoveReplica(rng1); err != nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc()); err != nil {
 		t.Error(err)
 	}
 	// Add 10 new ranges.
@@ -419,7 +446,7 @@ func TestStoreRangeSet(t *testing.T) {
 	if rng.Desc().RangeID != 2 {
 		t.Errorf("expected fetch of rangeID=2; got %d", rng.Desc().RangeID)
 	}
-	if err := store.RemoveReplica(rng); err != nil {
+	if err := store.RemoveReplica(rng, *rng.Desc()); err != nil {
 		t.Error(err)
 	}
 	if ec := ranges.EstimatedCount(); ec != 9 {
@@ -1528,7 +1555,7 @@ func TestMaybeRemove(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := store.RemoveReplica(rng); err != nil {
+	if err := store.RemoveReplica(rng, *rng.Desc()); err != nil {
 		t.Error(err)
 	}
 	// MaybeRemove is called.

